### PR TITLE
proc-cmdline: introduce PROC_CMDLINE_TRUE_WHEN_MISSING and use it where appropriate

### PR DESCRIPTION
**Pull request by YHNdnzj**
Split out from #28744

### CODE REVIEW

**Comment by mrc0mmand** [comment]
There's something very wrong going on with the C8S job in this PR (even after giving another attempt), I'll try to get some more info.

**Comment by YHNdnzj** [comment]
> There's something very wrong going on with the C8S job in this PR (even after giving itanother attempt), I'll try to get some more info.

Sorry, but it was my terribly foolish mistake... I hope it's fixed now.

File: src/shared/reboot-util.c [hunk 1]
```diff
@@ -110,14 +110,18 @@ int reboot_with_parameter(RebootFlags flags) {
...
+        bool b;
         int r;
 
-        r = proc_cmdline_get_bool("systemd.restore_state", /* flags = */ 0, &ret);
+        if (cached >= 0)
+                return cached;
+
+        r = proc_cmdline_get_bool("systemd.restore_state", PROC_CMDLINE_TRUE_WHEN_MISSING, &b);
         if (r < 0)
                 return r;
```
Comment (yuwata) [inline src/shared/reboot-util.c]: When you touch the code, please make the failure non-critical, like fstab-util or battery-check do.

File: src/shared/reboot-util.c [hunk 2]
```diff
@@ -110,14 +110,18 @@ int reboot_with_parameter(RebootFlags flags) {
 }
 
 int shall_restore_state(void) {
-        bool ret;
+        static int cached = -1;
+        bool b;
```
Comment (yuwata) [inline src/shared/reboot-util.c]: Then, initialize it with `true`, and please add a short comment.

**Review by yuwata — APPROVED** [review]
Thanks. Looks mostly good.

File: src/basic/proc-cmdline.h [hunk 3]
```diff
@@ -10,6 +10,7 @@ typedef enum ProcCmdlineFlags {
         PROC_CMDLINE_VALUE_OPTIONAL     = 1 << 1, /* the value is optional (for boolean switches that can omit the value) */
         PROC_CMDLINE_RD_STRICT          = 1 << 2, /* ignore this in the initrd */
         PROC_CMDLINE_IGNORE_EFI_OPTIONS = 1 << 3, /* don't check systemd's private EFI variable */
+        PROC_CMDLINE_TRUE_WHEN_MISSING  = 1 << 4, /* default to true when the key is missing for bool */
```
Comment (yuwata) [inline src/basic/proc-cmdline.h:13]: Could you add simple test cases for the flag?

**Review by yuwata — APPROVED** [review]

**Pull request merged**